cranelift/codegen/src/egraph/elaborate.rs

```diff
@@ -0,0 +1,589 @@
+                self.first_branch[to_block] = Some(inst).into();
+            }
+        } else if let Some(branch) = self.first_branch[to_block].into() {
+            self.func.layout.insert_inst(inst, branch);
+        } else {
+            self.func.layout.append_inst(inst, to_block);
+        }
+        self.func.dfg.inst_results_list(inst)
+    }
+
+    fn compute_best_nodes(&mut self) {
+        let best = &mut self.id_to_best_cost_and_node;
+        for (eclass_id, eclass) in &self.egraph.classes {
+            trace!("computing best for eclass {:?}", eclass_id);
+            if let Some(child1) = eclass.child1() {
+                trace!(" -> child {:?}", child1);
+                best[eclass_id] = best[child1];
+            }
+            if let Some(child2) = eclass.child2() {
+                trace!(" -> child {:?}", child2);
+                if best[child2].0 < best[eclass_id].0 {
+                    best[eclass_id] = best[child2];
+                }
+            }
+            if let Some(node_key) = eclass.get_node() {
+                let node = node_key.node(&self.egraph.nodes);
+                trace!(" -> eclass {:?}: node {:?}", eclass_id, node);
+                let (cost, id) = match node {
+                    Node::Param { .. }
+                    | Node::Inst { .. }
+                    | Node::Load { .. }
+                    | Node::Result { .. } => (Cost::zero(), eclass_id),
+                    Node::Pure { op, .. } => {
+                        let args_cost = self
+                            .node_ctx
+                            .children(node)
+                            .iter()
+                            .map(|&arg_id| {
+                                trace!("  -> arg {:?}", arg_id);
+                                best[arg_id].0
+                            })
+                            // Can't use `.sum()` for `Cost` types; do
+                            // an explicit reduce instead.
+                            .reduce(|a, b| a + b)
+                            .unwrap_or(Cost::zero());
+                        let level = self.egraph.analysis_value(eclass_id).loop_level;
+                        let cost = op_cost(op).at_level(level) + args_cost;
+                        (cost, eclass_id)
+                    }
+                };
+
+                if cost < best[eclass_id].0 {
+                    best[eclass_id] = (cost, id);
+                }
+            }
+            debug_assert_ne!(best[eclass_id].0, Cost::infinity());
+            debug_assert_ne!(best[eclass_id].1, Id::invalid());
+            trace!("best for eclass {:?}: {:?}", eclass_id, best[eclass_id]);
+        }
+    }
+
+    fn elaborate_eclass_use(&mut self, id: Id) -> IdValue {
+        self.elab_stack.push(ElabStackEntry::Start { id });
+        self.process_elab_stack();
+        debug_assert_eq!(self.elab_result_stack.len(), 1);
+        self.elab_result_stack.pop().expect("Must have one result")
```

Comment:
Ah, here we are actually returning the result, so unfortunately we can't do that...

Reply:
Err, apparently I'm reading too fast because I see you noted this above and further noted it isn't used :-) Yes, actually, this is a nice cleanup, thanks!